Prep input for differential expression

allyhawkins opened the pull request:
Closes #572 

Here I expanded on the snakefile that was set up in #580 to add in a rule that reads in all of the individually processed SCE objects, merges the objects into one combined object, identifies combined HVG's and integrates. The returned object is the merged object with the addition of the corrected PCs and UMAP calculated using `fastMNN`. This would allow us to read in a single object into the differential expression notebook that contains both the original expression dataset and the corrected PCA and UMAP. 

A few things to note: 

- I did not include the corrected gene expression data. I didn't do anything with it in the exploration notebook, but perhaps I should include it so that we could show why you wouldn't use that data for DE analysis? 
- I am using all 10 libraries here to create the integrated dataset, even though in the setup notebook we narrow it down to only 6 libraries that we actually use for DE. I'm torn on if we want to show some of the plots on how we narrowed it down to those 6 libraries and still start with the 10 libraries or just start with an object that has the 6 libraries. 
- I wrote the output into `rms/integrated` but named it `rms_all_integrated.rds` to hopefully not get confused with the output that will get created in the integration teaching notebook. Let me know if others have different ideas for organization.

jashapiro commented on scRNA-seq-advanced/setup/rms/integrate_rms.R:
```diff
@@ -0,0 +1,131 @@
+                              n = 2000)
+
+# Add PCA and UMAP into the SCE
+combined_sce <- scater::runPCA(combined_sce,
+                               subset_row = hvg_list) %>%
+  scater::runUMAP(dimred = "PCA")
```
Reformatting for a bit more clarity on the flow.
```suggestion
combined_sce <- combined_sce %>%
  scater::runPCA(subset_row = hvg_list) %>%
  scater::runUMAP(dimred = "PCA")
```

jashapiro commented on scRNA-seq-advanced/setup/rms/integrate_rms.R:
```diff
@@ -0,0 +1,131 @@
+# Integration ------------------------------------------------------------------
+
+# perform fastMNN integration 
+integrated_sce <- batchelor::fastMNN(combined_sce,
+                                     batch = combined_sce$sample)
```
Do we want to use `subset.row = hvg_list` here? That would better parallel what we have in the instruction notebook, if people were trying to recreate this object. If we do that, we will need to be sure to also include `correct.all = TRUE` to get out the full reconstructed matrix. 

 I do think it is worth adding the reconstructed gene expression  to the output, if only for future-proofing purposes. We _might_ want to look at it.

jashapiro reviewed (commented):
This looks good! I have some pretty small suggestions.

I do think that we should include the reconstructed/corrected gene expression data. We might not use it, but it might be nice to glance at for some visualizations, so having it in there isn't going to hurt much. Bigger object, but hopefully still not too huge.

I think starting with all 10 samples could be good. I like the idea of showing some of the QC/sample selection that went into choosing the six samples, and the easiest way may be to have them all in one object. Alternatively, that could just be explained in slides and then we read in only the six samples we are planning to use. 

I had a file name suggestion as well.

allyhawkins commented on scRNA-seq-advanced/setup/rms/Snakefile:
```diff
@@ -28,4 +29,18 @@ rule prepare_library:
+  shell:
+    """
+      Rscript integrate_rms.R \
+        --input_file_list "{input}" \
```
Not quite comma separated list, it's a space-separated list though so still parsable!

allyhawkins commented on scRNA-seq-advanced/setup/rms/integrate_rms.R:
```diff
@@ -0,0 +1,131 @@
+# load training modules R proj
+training_modules_dir <- file.path(here::here(), "..")
+renv::load(training_modules_dir)
```
Okay, I added this because I was getting errors that `batchelor` wasn't installed even though it's in the lock file, but I can remove this with the assumption that we would have it installed.

jashapiro replied:
yeah, you may have to install it yourself for now, but we will make sure everybody has it.

jashapiro commented on scRNA-seq-advanced/setup/rms/integrate_rms.R:
```diff
@@ -0,0 +1,131 @@
+options <- list(
+  make_option(
+    "--input_file_list",
+    help = "Comma separated list of processed SCE objects to integrate."
```
Given [prior comment](https://github.com/AlexsLemonade/training-modules/pull/589/files#r1047602828)
```suggestion
    help = "Space separated list of processed SCE objects to integrate."
```

allyhawkins commented:
Thanks for looking at this @jashapiro. I updated the file name like you suggested, and passed on those changes to the syncing scripts. I also noticed that I hadn't synced the sample metadata that includes the diagnoses for each sample, so I added that to the scripts as well. 

The corrected gene expression data is now returned in the object and I also added the list of HVG's used to the metadata so that we can use those later if we need to. The other objects I had been working with only returned the HVG's and didn't include all genes so we might want to do that as well before doing any DE. This should be ready for another look.

jashapiro reviewed (approved):
This looks good to me, though I have not tried to run it myself.

Before merging, I delete the contents of `{basedir}/integrated/` and run it one more time, just to be sure that everything works as expected and there are no extra files there.

Then run `link-data.sh` and `syncup-s3.sh`. I believe you should have all permissions that you need, but let me know if your run into any permissions issues with those. Do not be too surprised if you see some additional files syncing up to S3... as you know, we can't always predict what `aws s3 sync` will think is new!
